File: app/src/main/java/com/nfcalarmclock/NacNotification.java
package com.nfcalarmclock;

import android.annotation.TargetApi;
import android.app.Notification;
import android.app.NotificationChannel;
import android.app.PendingIntent;
import android.content.Context;
import android.os.Build;
import androidx.core.app.NotificationCompat;
import androidx.core.app.NotificationManagerCompat;
import java.util.List;
import java.util.Locale;

/**
 * Notification for the app to keep it in memory.
 */
public abstract class NacNotification
{

	/**
	 * @return Channel description.
	 */
	protected abstract String getChannelDescription();

	/**
	 * @return Channel ID.
	 */
	protected abstract String getChannelId();

	/**
	 * @return Channel name.
	 */
	protected abstract String getChannelName();

	/**
	 * @return Content pending intent when building the notification.
	 */
	protected abstract PendingIntent getContentPendingIntent();

	/**
	 * @return Content text when building the notification.
	 */
	protected abstract String getContentText();

	/**
	 * @return Group name.
	 */
	protected abstract String getGroup();

	/**
	 * @return Notification ID.
	 */
	protected abstract int getId();

	/**
	 * @return Level of importance of the notification.
	 */
	protected abstract int getImportance();

	/**
	 * @return Level of priority of the notification.
	 */
	protected abstract int getPriority();

	/**
	 * @return Notification title.
	 */
	public abstract String getTitle();

	/**
	 * Context.
	 */
	protected final Context mContext;

	/**
	 * Body text of the notification.
	 */
	protected List<CharSequence> mBody;

	/**
	 */
	public NacNotification()
	{
		super();
		this.mContext = null;
		this.mBody = null;
	}

	/**
	 */
	public NacNotification(Context context)
	{
		this.mContext = context;
		this.mBody = null;
		this.setupChannel();
	}

	/**
	 * Build the notification.
	 */
	@SuppressWarnings("deprecation")
	@TargetApi(Build.VERSION_CODES.N)
	protected NotificationCompat.Builder builder()
	{
		return new NotificationCompat.Builder(this.getContext(), this.getChannelId())
			.setGroup(this.getGroup())
			.setContentIntent(this.getContentPendingIntent())
			.setContentTitle(NacUtility.toSpannedString(this.getTitle()))
			.setContentText(this.getContentText())
			.setSmallIcon(this.getSmallIcon())
			.setPriority(this.getPriority())
			.setCategory(this.getCategory());
	}

	/**
	 * Cancel a previously shown notification.
	 */
	public void cancel()
	{
		NotificationManagerCompat manager = this.getNotificationManager();
		int id = this.getId();
		manager.cancel(id);
	}

	/**
	 * Create the notification channel.
	 */
	@TargetApi(Build.VERSION_CODES.O)
	protected NotificationChannel createChannel()
	{
		if (Build.VERSION.SDK_INT < Build.VERSION_CODES.O)
		{
			return null;
		}

		NotificationChannel channel = new NotificationChannel(
			this.getChannelId(),
			this.getChannelName(),
			this.getImportance());

		channel.setDescription(this.getChannelDescription());
		channel.setLockscreenVisibility(Notification.VISIBILITY_PRIVATE);
		return channel;
	}

	/**
	 * @return True if the given group key matches that of the notification, and
	 *         False otherwise.
	 */
	protected boolean doesGroupMatch(String groupKey)
	{
		return this.getGroup().equals(groupKey);
	}

	/**
	 * @return Body text of the notification.
	 */
	protected List<CharSequence> getBody()
	{
		return this.mBody;
	}

	/**
	 * @return A line that can appear in the body text of the notification.
	 */
	protected String getBodyLine(NacAlarm alarm)
	{
		Context context = this.getContext();
		String time = alarm.getFullTime(context);
		String name = alarm.getName();
		Locale locale = Locale.getDefault();

		return name.isEmpty() ? time
			: String.format(locale, "%1$s  —  %2$s", time, name);
	}

	/**
	 * @return The category of the notification.
	 */
	protected String getCategory()
	{
		return NotificationCompat.CATEGORY_ALARM;
	}

	/**
	 * @return The context.
	 */
	protected Context getContext()
	{
		return this.mContext;
	}

	/**
	 * @return Number of lines in the body.
	 * 
	 * TODO: See if this can be 0 instead of 1.
	 */
	protected int getLineCount()
	{
		List<CharSequence> body = this.getBody();
		return (body == null) ? 1 : body.size();
	}

	/**
	 * @return The notification manager.
	 */
	protected NotificationManagerCompat getNotificationManager()
	{
		Context context = this.getContext();
		return NotificationManagerCompat.from(context);
	}

	/**
	 * @return The small icon of the notification.
	 */
	protected int getSmallIcon()
	{
		return R.mipmap.notification;
	}

	/**
	 * @return True if there is text in the body, and False otherwise.
	 */
	protected boolean hasBody()
	{
		List<CharSequence> body = this.getBody();
		int size = (body != null) ? body.size() : 0;
		return (size > 0);
	}

	/**
	 * Setup the notification body lines.
	 */
	protected void setupBody()
	{
	}

	/**
	 * Setup the notification channel.
	 */
	protected void setupChannel()
	{
		Context context = this.getContext();
		NotificationChannel channel = this.createChannel();
		if (channel != null)
		{
			NotificationManagerCompat manager = this.getNotificationManager();
			manager.createNotificationChannel(channel);
		}
	}

	/**
	 * Show the notification.
	 */
	protected void show()
	{
		NotificationCompat.Builder builder = this.builder();
		NotificationManagerCompat manager = this.getNotificationManager();
		int id = this.getId();

		manager.notify(id, builder.build());
	}

}

File: app/src/main/java/com/nfcalarmclock/view/notification/NacNotification.kt
package com.nfcalarmclock.view.notification

import android.Manifest
import android.annotation.TargetApi
import android.app.Notification
import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.PendingIntent
import android.content.Context
import android.content.pm.PackageManager
import android.os.Build
import androidx.core.app.ActivityCompat
import androidx.core.app.NotificationCompat
import androidx.core.app.NotificationManagerCompat
import com.nfcalarmclock.R
import com.nfcalarmclock.alarm.db.NacAlarm
import com.nfcalarmclock.util.NacUtility.toSpannedString
import java.util.Arrays
import java.util.Locale

/**
 * Notification for the app to keep it in memory.
 */
abstract class NacNotification(

	/**
	 * Context.
	 */
	protected val context: Context
)
{

	/**
	 * Notification ID.
	 */
	protected abstract val id: Int

	/**
	 * Channel ID.
	 */
	protected abstract val channelId: String

	/**
	 * Channel name.
	 */
	protected abstract val channelName: String

	/**
	 * Channel description.
	 */
	protected abstract val channelDescription: String

	/**
	 * Notification title.
	 */
	abstract val title: String

	/**
	 * Priority level of the notification.
	 */
	protected abstract val priority: Int

	/**
	 * Importance level of the notification.
	 */
	protected abstract val importance: Int

	/**
	 * Group name.
	 */
	protected abstract val group: String

	/**
	 * Context text when building the notification.
	 */
	protected abstract val contentText: String

	/**
	 * Context pending intent when building the notification.
	 */
	protected abstract val contentPendingIntent: PendingIntent

	/**
	 * The notification manager.
	 */
	private val notificationManager: NotificationManagerCompat
		get() = NotificationManagerCompat.from(context)

	/**
	 * Category of the notification.
	 */
	private val category: String = NotificationCompat.CATEGORY_ALARM

	/**
	 * Body text of the notification.
	 */
	protected var body: List<CharSequence> = emptyList()

	/**
	 * The small icon of the notification.
	 */
	private val smallIcon: Int
		get() = R.mipmap.notification

	/**
	 * Get the number of lines in the body.
	 *
	 * TODO: See if this can be 0 instead of 1.
	 *
	 * @return Number of lines in the body.
	 */
	protected val lineCount: Int
		get() = body.size

	/**
	 * Check if should create channel by checkin if the API level < 26.
	 */
	private val shouldCreateChannel: Boolean
		get() = Build.VERSION.SDK_INT >= Build.VERSION_CODES.O

	/**
	 * Constructor.
	 */
	init
	{
		setupChannel()
	}

	/**
	 * Build the notification.
	 */
	@TargetApi(Build.VERSION_CODES.N)
	protected open fun builder(): NotificationCompat.Builder
	{
		return NotificationCompat.Builder(context, channelId)
			.setGroup(group)
			.setContentIntent(contentPendingIntent)
			.setContentTitle(toSpannedString(title))
			.setContentText(contentText)
			.setSmallIcon(smallIcon)
			.setPriority(priority)
			.setCategory(category)
	}

	/**
	 * Cancel a previously shown notification.
	 */
	fun cancel()
	{
		notificationManager.cancel(id)
	}

	/**
	 * Create the notification channel.
	 */
	@TargetApi(Build.VERSION_CODES.O)
	protected open fun createChannel(): NotificationChannel
	{
		// Create the channel
		val channel = NotificationChannel(channelId, channelName, importance)

		// Basic setup of the channel
		channel.description = channelDescription
		channel.lockscreenVisibility = Notification.VISIBILITY_PRIVATE

		return channel
	}

	/**
	 * Check if the given group key matches that of the notification.
	 *
	 * @return True if the given group key matches that of the notification, and
	 *         False otherwise.
	 */
	@Suppress("unused")
	protected fun doesGroupMatch(groupKey: String): Boolean
	{
		return group == groupKey
	}

	/**
	 * Get a line that can appear in the body text of the notification.
	 *
	 * @return A line that can appear in the body text of the notification.
	 */
	protected fun getBodyLine(alarm: NacAlarm): String
	{
		val time = alarm.getFullTime(context)
		val name = alarm.name
		val locale = Locale.getDefault()

		// Check if the alarm name is empty
		return if (name.isEmpty())
			{
				// The line in the notification will just be the time
				time
			}
			else
			{
				// The line in the notification will be the time and the name of the
				// alarm
				String.format(locale, "%1\$s  —  %2\$s", time, name)
			}
	}

	/**
	 * Setup the notification body lines.
	 */
	@Suppress("unused")
	protected open fun setupBody()
	{
	}

	/**
	 * Setup the notification channel.
	 */
	private fun setupChannel()
	{
		// Check if should create channel
		if (!shouldCreateChannel)
		{
			println("Will not create channel")
			return
		}

		// Create the channel
		println("Create channel")
		val channel = createChannel()

		// Tell the notification manager to create the channel
		notificationManager.createNotificationChannel(channel)
	}

	/**
	 * Show the notification.
	 */
	protected open fun show()
	{
		// Build the notification
		val builder = builder()

		// Check that the app is able to post notification on APK >= 33
		if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU)
		{
			if (ActivityCompat.checkSelfPermission(context, Manifest.permission.POST_NOTIFICATIONS)
				!= PackageManager.PERMISSION_GRANTED)
			{
				// TODO: Consider calling
				//    ActivityCompat#requestPermissions
				// here to request the missing permissions, and then overriding
				//   public void onRequestPermissionsResult(int requestCode, String[] permissions,
				//                                          int[] grantResults)
				// to handle the case where the user grants the permission. See the documentation
				// for ActivityCompat#requestPermissions for more details.
				return
			}
		}

		// Show the notification
		notificationManager.notify(id, builder.build())
	}

	companion object
	{

		/**
		 * Get a list of notification lines.
		 *
		 * @return A list of notification lines.
		 */
		@TargetApi(Build.VERSION_CODES.M)
		fun getExtraLines(context: Context, groupKey: String?): MutableList<CharSequence>
		{
			// Create a list of lines
			val lines: MutableList<CharSequence> = ArrayList()

			// Check if API < 23
			if (Build.VERSION.SDK_INT < Build.VERSION_CODES.M)
			{
				return lines
			}

			// Get the notification manager
			val manager = context.getSystemService(Context.NOTIFICATION_SERVICE)
				as NotificationManager

			// Get the active notifications in the status bar
			val statusbar = manager.activeNotifications

			// Iterate over each notification
			for (sb in statusbar)
			{
				val notification = sb.notification
				val sbGroup = notification.group

				// Check that the groups match
				if ((groupKey != null) && (groupKey == sbGroup))
				{
					val extraLines = notification.extras.getCharSequenceArray(
						NotificationCompat.EXTRA_TEXT_LINES)

					// Check that the extra lines are not null
					if (extraLines != null)
					{
						lines.addAll(Arrays.asList(*extraLines))
					}
				}
			}

			return lines
		}

	}

}